ballerina/client_oauth2_provider.bal

```diff
@@ -416,39 +428,52 @@ isolated function getAccessTokenFromRefreshRequestForPasswordGrant(PasswordGrant
     var refreshConfig = config?.refreshConfig;
     if refreshConfig is () {
         return prepareError("Failed to refresh access token since refresh configurations are not provided.");
-    } else {
-        string? clientId = config?.clientId;
-        string? clientSecret = config?.clientSecret;
-        if clientId is string && clientSecret is string {
-            // Checking `(clientId == "" || clientSecret == "")` is validated while requesting access token by token
-            // request, initially.
-            string refreshUrl = refreshConfig.refreshUrl;
-            string refreshToken = tokenCache.getRefreshToken();
-            if refreshToken == "" {
-                // The subsequent requests should have a cached `refreshToken` to refresh the access token.
-                return prepareError("Failed to refresh access token since refresh-token has not been cached from the initial authorization response.");
-            }
-            RequestConfig requestConfig = {
-                payload: "grant_type=refresh_token&refresh_token=" + refreshToken,
-                clientId: clientId,
-                clientSecret: clientSecret,
-                scopes: refreshConfig?.scopes,
-                optionalParams: refreshConfig?.optionalParams,
-                credentialBearer: refreshConfig.credentialBearer
-            };
-            ClientConfiguration clientConfig = refreshConfig.clientConfig;
-            decimal defaultTokenExpTime = config.defaultTokenExpTime;
-            decimal clockSkew = config.clockSkew;
-
-            json response = check sendRequest(requestConfig, refreshUrl, clientConfig);
-            string accessToken = check extractAccessToken(response);
-            string? updatedRefreshToken = extractRefreshToken(response);
-            int? expiresIn = extractExpiresIn(response);
-            tokenCache.update(accessToken, updatedRefreshToken, expiresIn, defaultTokenExpTime, clockSkew);
-            return accessToken;
+    }
+
+    if refreshConfig is INFER_REFRESH_CONFIG {
+        refreshConfig = {
+            refreshUrl: config.tokenUrl,
+            optionalParams: config.optionalParams,
+            credentialBearer: config.credentialBearer,
+            clientConfig: config.clientConfig
+        };
+    }
+
+    if refreshConfig !is RefreshConfig {
+        return prepareError("Invalid refresh configuration. Failed to refresh access token.");
+    }
```

Comment:
Do we need this validation? The type of the `refreshConfig` should be `RefreshConfig|INFER_REFRESH_CONFIG?` right? In that case, this seems like unreachable code (Maybe the compiler does not complain about that). But I think we can simplify this to avoid this additional check.

This function also seems to be doing more than one thing.
- Retrieving the `refreshConfig` from the `config`
- Validating the retrieved `refreshConfig`
- Retrieving the access token from the refresh request

We can refactor this and break it into separate functions.

Reply:
+1 for refactoring the code

Reply:
Fixed with [244291e](https://github.com/ballerina-platform/module-ballerina-oauth2/pull/869/commits/244291e4362cfff3dfc58a9ec483857e4c507752)